# Exploratory Data Analysis

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
file_path = "./DataSets/"
training_data = pd.read_csv(file_path + 'counterfeit_train.csv')
production_data = pd.read_csv(file_path + 'counterfeit_test.csv')

In [3]:
training_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  Medicine_MRP  \
0       RRA15              13.100     Area046          1995      160.2366   
1       YVV26                 NaN     Area027          1983      110.4384   
2       LJC15               9.025     Area046          1995      259.4092   
3       GWC40              11.800     Area046          1995       99.9830   
4       QMN13                 NaN     Area019          1983       56.4402   

        Medicine_Type SidEffect_Level  Availability_rating  \
0        Antimalarial        critical             0.070422   
1         Mstablizers            mild             0.013000   
2             Cardiac            mild             0.060783   
3  OralContraceptives            mild             0.065555   
4       Hreplacements        critical             0.248859   

            Area_Type Area_City_Type Area_dist_level  Counterfeit_Sales  
0            DownTown         Tier 1           Small          1775.5026  
1          CityLimits         Tier 3          Medium          3069.1520  
2            DownTown         Tier 1           Small          2603.0920  
3            DownTown         Tier 1           Small          1101.7130  
4  MidTownResidential         Tier 1           Small           158.9402

In [4]:
production_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  Medicine_MRP  \
0       HLZ81                 NaN     Area027          1983       85.5328   
1       ECE94               13.45     Area045          2000      257.1460   
2       SAD14                7.10     Area045          2000       98.1172   
3       EQV63               18.30     Area010          1996      135.3730   
4       AIR10                 NaN     Area019          1983      112.8016   

        Medicine_Type SidEffect_Level  Availability_rating  \
0         Antibiotics            mild             0.112747   
1  OralContraceptives            mild             0.144446   
2        Antipyretics            mild             0.144221   
3       Tranquilizers            mild             0.100388   
4  OralContraceptives            mild             0.022585   

            Area_Type Area_City_Type Area_dist_level  
0          CityLimits         Tier 3          Medium  
1            DownTown         Tier 2         Unknown  
2            DownTown         Tier 2         Unknown  
3  MidTownResidential         Tier 3         Unknown  
4  MidTownResidential         Tier 1           Small

In [5]:
training_data['Counterfeit_Weight'].mean()

14.115056617126996

In [6]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6818 entries, 0 to 6817
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Medicine_ID          6818 non-null   object 
 1   Counterfeit_Weight   5652 non-null   float64
 2   DistArea_ID          6818 non-null   object 
 3   Active_Since         6818 non-null   int64  
 4   Medicine_MRP         6818 non-null   float64
 5   Medicine_Type        6818 non-null   object 
 6   SidEffect_Level      6818 non-null   object 
 7   Availability_rating  6818 non-null   float64
 8   Area_Type            6818 non-null   object 
 9   Area_City_Type       6818 non-null   object 
 10  Area_dist_level      6818 non-null   object 
 11  Counterfeit_Sales    6818 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 639.3+ KB


In [7]:
production_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1705 entries, 0 to 1704
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Medicine_ID          1705 non-null   object 
 1   Counterfeit_Weight   1408 non-null   float64
 2   DistArea_ID          1705 non-null   object 
 3   Active_Since         1705 non-null   int64  
 4   Medicine_MRP         1705 non-null   float64
 5   Medicine_Type        1705 non-null   object 
 6   SidEffect_Level      1705 non-null   object 
 7   Availability_rating  1705 non-null   float64
 8   Area_Type            1705 non-null   object 
 9   Area_City_Type       1705 non-null   object 
 10  Area_dist_level      1705 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 146.6+ KB


Creating a Dummy sales variable in the production data and combining it with training data for further data manipulations

In [8]:
production_data['Counterfeit_Sales'] = np.NAN

In [9]:
training_data['Data_from'] = 'Train'
production_data['Data_from'] = 'Production'

In [10]:
full_data = pd.concat([training_data,production_data],axis=0)

In [11]:
full_data.loc[full_data['Data_from'] == 'Train',]

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  Medicine_MRP  \
0          RRA15              13.100     Area046          1995      160.2366   
1          YVV26                 NaN     Area027          1983      110.4384   
2          LJC15               9.025     Area046          1995      259.4092   
3          GWC40              11.800     Area046          1995       99.9830   
4          QMN13                 NaN     Area019          1983       56.4402   
...          ...                 ...         ...           ...           ...   
6813       OYN80               8.535     Area046          1995      204.1452   
6814       ACW12              20.650     Area046          1995      235.1088   
6815       OPM10              20.000     Area017          2005      193.6292   
6816       SLY12              10.180     Area045          2000      162.8682   
6817       ATT10               7.930     Area018          2007      177.5500   

           Medicine_Type SidEffect_Level  Availability_rating  \
0           Antimalarial        critical             0.070422   
1            Mstablizers            mild             0.013000   
2                Cardiac            mild             0.060783   
3     OralContraceptives            mild             0.065555   
4          Hreplacements        critical             0.248859   
...                  ...             ...                  ...   
6813       Hreplacements            mild             0.112963   
6814       Hreplacements            mild             0.131103   
6815        Antimalarial        critical             0.105096   
6816             Statins            mild             0.099957   
6817         Mstablizers        critical             0.117445   

               Area_Type Area_City_Type Area_dist_level  Counterfeit_Sales  \
0               DownTown         Tier 1           Small          1775.5026   
1             CityLimits         Tier 3          Medium          3069.1520   
2               DownTown         Tier 1           Small          2603.0920   
3               DownTown         Tier 1           Small          1101.7130   
4     MidTownResidential         Tier 1           Small           158.9402   
...                  ...            ...             ...                ...   
6813            DownTown         Tier 1           Small          2070.4520   
6814            DownTown         Tier 1           Small          2126.3792   
6815            DownTown         Tier 2         Unknown          2119.7212   
6816            DownTown         Tier 2         Unknown          1485.2138   
6817          Industrial         Tier 3          Medium          1943.9500   

     Data_from  
0        Train  
1        Train  
2        Train  
3        Train  
4        Train  
...        ...  
6813     Train  
6814     Train  
6815     Train  
6816     Train  
6817     Train  

[6818 rows x 13 columns]

In [12]:
full_data['DistArea_ID'].value_counts()

Area027    935
Area013    932
Area035    930
Area046    930
Area049    930
Area045    929
Area018    928
Area017    926
Area010    555
Area019    528
Name: DistArea_ID, dtype: int64

In [13]:
full_data['Area_Type'].value_counts()

DownTown              5577
MidTownResidential    1083
CityLimits             935
Industrial             928
Name: Area_Type, dtype: int64

In [14]:
full_data.loc[full_data['Area_Type']=='DownTown','DistArea_ID'].unique()

array(['Area046', 'Area045', 'Area013', 'Area017', 'Area035', 'Area049'],
      dtype=object)

In [15]:
full_data.loc[full_data['Area_Type']=='MidTownResidential','DistArea_ID'].unique()

array(['Area019', 'Area010'], dtype=object)

In [16]:
full_data.loc[full_data['Area_Type']=='CityLimits','DistArea_ID'].unique()

array(['Area027'], dtype=object)

In [17]:
full_data.loc[full_data['Area_Type']=='Industrial','DistArea_ID'].unique()

array(['Area018'], dtype=object)

In [18]:
full_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  Medicine_MRP  \
0       RRA15              13.100     Area046          1995      160.2366   
1       YVV26                 NaN     Area027          1983      110.4384   
2       LJC15               9.025     Area046          1995      259.4092   
3       GWC40              11.800     Area046          1995       99.9830   
4       QMN13                 NaN     Area019          1983       56.4402   

        Medicine_Type SidEffect_Level  Availability_rating  \
0        Antimalarial        critical             0.070422   
1         Mstablizers            mild             0.013000   
2             Cardiac            mild             0.060783   
3  OralContraceptives            mild             0.065555   
4       Hreplacements        critical             0.248859   

            Area_Type Area_City_Type Area_dist_level  Counterfeit_Sales  \
0            DownTown         Tier 1           Small          1775.5026   
1          CityLimits         Tier 3          Medium          3069.1520   
2            DownTown         Tier 1           Small          2603.0920   
3            DownTown         Tier 1           Small          1101.7130   
4  MidTownResidential         Tier 1           Small           158.9402   

  Data_from  
0     Train  
1     Train  
2     Train  
3     Train  
4     Train

In [19]:
full_data['Medicine_Type'].value_counts()

Antibiotics           1232
Hreplacements         1200
Antiseptics            910
OralContraceptives     856
Antipyretics           682
Cardiac                649
Mstablizers            648
Tranquilizers          520
Analgesics             445
Antimalarial           425
Antacids               251
Statins                214
MuscleRelaxants        169
Antifungal             148
Stimulants             110
Antiviral               64
Name: Medicine_Type, dtype: int64

In [20]:
round(full_data.groupby('Medicine_Type')['Counterfeit_Sales'].mean(),2)

Medicine_Type
Analgesics            2091.24
Antacids              2259.88
Antibiotics           2397.20
Antifungal            2444.06
Antimalarial          2192.68
Antipyretics          2292.72
Antiseptics           2366.59
Antiviral             2583.16
Cardiac               2362.98
Hreplacements         2376.12
Mstablizers           2044.87
MuscleRelaxants       2091.09
OralContraceptives    2245.22
Statins               2309.45
Stimulants            2279.97
Tranquilizers         2084.45
Name: Counterfeit_Sales, dtype: float64

In [21]:
full_data['SidEffect_Level'].value_counts()

mild        5517
critical    3006
Name: SidEffect_Level, dtype: int64

In [22]:
full_data['Area_dist_level'].value_counts()

Medium     2793
Unknown    2410
Small      2388
High        932
Name: Area_dist_level, dtype: int64

#### After looking into all the categorical  variables, we cannot drop them and we create dummies for each variable for simplicity later we can ckeck for additional options to improve the performance

In [23]:
full_data.head(10)

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  Medicine_MRP  \
0       RRA15              13.100     Area046          1995      160.2366   
1       YVV26                 NaN     Area027          1983      110.4384   
2       LJC15               9.025     Area046          1995      259.4092   
3       GWC40              11.800     Area046          1995       99.9830   
4       QMN13                 NaN     Area019          1983       56.4402   
5       JDG81               8.775     Area045          2000      165.5656   
6       KPX48              18.000     Area018          2007      156.9102   
7       CYW14               7.680     Area013          1985      154.3470   
8       TVF57              18.550     Area046          1995      107.3068   
9       MGU75               9.510     Area013          1985      161.1392   

        Medicine_Type SidEffect_Level  Availability_rating  \
0        Antimalarial        critical             0.070422   
1         Mstablizers            mild             0.013000   
2             Cardiac            mild             0.060783   
3  OralContraceptives            mild             0.065555   
4       Hreplacements        critical             0.248859   
5         Antiseptics            mild             0.088881   
6       Hreplacements        critical             0.051685   
7         Antiseptics            mild             0.028150   
8       Hreplacements        critical             0.060266   
9             Statins            mild             0.023755   

            Area_Type Area_City_Type Area_dist_level  Counterfeit_Sales  \
0            DownTown         Tier 1           Small          1775.5026   
1          CityLimits         Tier 3          Medium          3069.1520   
2            DownTown         Tier 1           Small          2603.0920   
3            DownTown         Tier 1           Small          1101.7130   
4  MidTownResidential         Tier 1           Small           158.9402   
5            DownTown         Tier 2         Unknown          3047.8464   
6          Industrial         Tier 3          Medium          2883.3938   
7            DownTown         Tier 3            High          3262.2340   
8            DownTown         Tier 1           Small          1959.9292   
9            DownTown         Tier 3            High          3095.7840   

  Data_from  
0     Train  
1     Train  
2     Train  
3     Train  
4     Train  
5     Train  
6     Train  
7     Train  
8     Train  
9     Train

In [24]:
d = pd.get_dummies(full_data['DistArea_ID'],prefix='DistID',drop_first=True)

In [25]:
full_data = pd.concat((full_data,d),axis='columns')

In [26]:
full_data.drop('DistArea_ID',axis='columns',inplace=True)

In [27]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

           Medicine_Type SidEffect_Level  Availability_rating  \
0           Antimalarial        critical             0.070422   
1            Mstablizers            mild             0.013000   
2                Cardiac            mild             0.060783   
3     OralContraceptives            mild             0.065555   
4          Hreplacements        critical             0.248859   
...                  ...             ...                  ...   
1700       Hreplacements            mild             0.050505   
1701         Antibiotics        critical             0.041118   
1702         Antiseptics            mild             0.191273   
1703       Hreplacements            mild             0.013000   
1704         Antibiotics        critical             0.098982   

               Area_Type Area_City_Type Area_dist_level  ...   Data_from  \
0               DownTown         Tier 1           Small  ...       Train   
1             CityLimits         Tier 3          Medium  ...       Train   
2               DownTown         Tier 1           Small  ...       Train   
3               DownTown         Tier 1           Small  ...       Train   
4     MidTownResidential         Tier 1           Small  ...       Train   
...                  ...            ...             ...  ...         ...   
1700          CityLimits         Tier 3          Medium  ...  Production   
1701            DownTown         Tier 2           Small  ...  Production   
1702            DownTown         Tier 2         Unknown  ...  Production   
1703          Industrial         Tier 3          Medium  ...  Production   
1704            DownTown         Tier 2         Unknown  ...  Production   

     DistID_Area013  DistID_Area017  DistID_Area018  DistID_Area019  \
0                 0               0               0               0   
1                 0               0               0               0   
2                 0               0               0               0   
3                 0               0               0               0   
4                 0               0               0               1   
...             ...             ...             ...             ...   
1700              0               0               0               0   
1701              0               0               0               0   
1702              0               1               0               0   
1703              0               0               1               0   
1704              0               0               0               0   

      DistID_Area027  DistID_Area035  DistID_Area045  DistID_Area046  \
0                  0               0               0               1   
1                  1               0               0               0   
2                  0               0               0               1   
3                  0               0               0               1   
4                  0               0               0               0   
...              ...             ...             ...             ...   
1700               1               0               0               0   
1701               0               1               0             

##### Medicine variable - Dummy Creations

In [28]:
d = pd.get_dummies(full_data['Medicine_Type'],prefix='MedType',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Medicine_Type',axis='columns',inplace=True,)

In [29]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

     SidEffect_Level  Availability_rating           Area_Type Area_City_Type  \
0           critical             0.070422            DownTown         Tier 1   
1               mild             0.013000          CityLimits         Tier 3   
2               mild             0.060783            DownTown         Tier 1   
3               mild             0.065555            DownTown         Tier 1   
4           critical             0.248859  MidTownResidential         Tier 1   
...              ...                  ...                 ...            ...   
1700            mild             0.050505          CityLimits         Tier 3   
1701        critical             0.041118            DownTown         Tier 2   
1702            mild             0.191273            DownTown         Tier 2   
1703            mild             0.013000          Industrial         Tier 3   
1704        critical             0.098982            DownTown         Tier 2   

     Area_dist_level  Counterfeit_Sales  ... MedType_Antiseptics  \
0              Small          1775.5026  ...                   0   
1             Medium          3069.1520  ...                   0   
2              Small          2603.0920  ...                   0   
3              Small          1101.7130  ...                   0   
4              Small           158.9402  ...                   0   
...              ...                ...  ...                 ...   
1700          Medium                NaN  ...                   0   
1701           Small                NaN  ...                   0   
1702         Unknown                NaN  ...                   1   
1703          Medium                NaN  ...                   0   
1704         Unknown                NaN  ...                   0   

      MedType_Antiviral  MedType_Cardiac  MedType_Hreplacements  \
0                     0                0                      0   
1                     0                0                      0   
2                     0                1                      0   
3                     0                0                      0   
4                     0                0                      1   
...                 ...              ...                    ...   
1700                  0                0                      1   
1701                  0                0                      0   
1702                  0                0                      0   
1703                  0                0                      1   
1704                  0                0                      0   

      MedType_Mstablizers  MedType_MuscleRelaxants  \
0                       0                        0   
1                       1                        0   
2                       0                        0   
3                       0                        0   
4                       0                        0   
...                   ...                      ...   
1700                    0                        0   
1701                    0                        0   
1702                    0                        0   
1703                    0                        0   
1704       

##### SidEffect_Level - Dummy Creation

In [30]:
full_data['SidEffect_Level'].value_counts()

mild        5517
critical    3006
Name: SidEffect_Level, dtype: int64

In [31]:
d = pd.get_dummies(full_data['SidEffect_Level'],prefix='SidLvl',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('SidEffect_Level',axis='columns',inplace = True)

In [32]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

      Availability_rating           Area_Type Area_City_Type Area_dist_level  \
0                0.070422            DownTown         Tier 1           Small   
1                0.013000          CityLimits         Tier 3          Medium   
2                0.060783            DownTown         Tier 1           Small   
3                0.065555            DownTown         Tier 1           Small   
4                0.248859  MidTownResidential         Tier 1           Small   
...                   ...                 ...            ...             ...   
1700             0.050505          CityLimits         Tier 3          Medium   
1701             0.041118            DownTown         Tier 2           Small   
1702             0.191273            DownTown         Tier 2         Unknown   
1703             0.013000          Industrial         Tier 3          Medium   
1704             0.098982            DownTown         Tier 2         Unknown   

      Counterfeit_Sales   Data_from  ...  MedType_Antiviral  MedType_Cardiac  \
0             1775.5026       Train  ...                  0                0   
1             3069.1520       Train  ...                  0                0   
2             2603.0920       Train  ...                  0                1   
3             1101.7130       Train  ...                  0                0   
4              158.9402       Train  ...                  0                0   
...                 ...         ...  ...                ...              ...   
1700                NaN  Production  ...                  0                0   
1701                NaN  Production  ...                  0                0   
1702                NaN  Production  ...                  0                0   
1703                NaN  Production  ...                  0                0   
1704                NaN  Production  ...                  0                0   

      MedType_Hreplacements  MedType_Mstablizers  MedType_MuscleRelaxants  \
0                         0                    0                        0   
1                         0                    1                        0   
2                         0                    0                        0   
3                         0                    0                        0   
4                         1                    0                        0   
...                     ...                  ...                      ...   
1700                      1                    0                        0   
1701                      0                    0                        0   
1702                      0                    0                        0   
1703                      1                    0                        0   
1704                      0                    0                        0   

      MedType_OralContraceptives  MedType_Statins  MedType_Stimulants  \
0                              0                0                   0   
1                              0                0                   0   
2                              0                0                   0   
3                              1                0

##### Area_Type - Dummy Creation

In [33]:
d = pd.get_dummies(full_data['Area_Type'],prefix='ArTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_Type',axis='columns',inplace = True)

In [34]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

      Availability_rating Area_City_Type Area_dist_level  Counterfeit_Sales  \
0                0.070422         Tier 1           Small          1775.5026   
1                0.013000         Tier 3          Medium          3069.1520   
2                0.060783         Tier 1           Small          2603.0920   
3                0.065555         Tier 1           Small          1101.7130   
4                0.248859         Tier 1           Small           158.9402   
...                   ...            ...             ...                ...   
1700             0.050505         Tier 3          Medium                NaN   
1701             0.041118         Tier 2           Small                NaN   
1702             0.191273         Tier 2         Unknown                NaN   
1703             0.013000         Tier 3          Medium                NaN   
1704             0.098982         Tier 2         Unknown                NaN   

       Data_from  DistID_Area013  ...  MedType_Mstablizers  \
0          Train               0  ...                    0   
1          Train               0  ...                    1   
2          Train               0  ...                    0   
3          Train               0  ...                    0   
4          Train               0  ...                    0   
...          ...             ...  ...                  ...   
1700  Production               0  ...                    0   
1701  Production               0  ...                    0   
1702  Production               0  ...                    0   
1703  Production               0  ...                    0   
1704  Production               0  ...                    0   

      MedType_MuscleRelaxants  MedType_OralContraceptives  MedType_Statins  \
0                           0                           0                0   
1                           0                           0                0   
2                           0                           0                0   
3                           0                           1                0   
4                           0                           0                0   
...                       ...                         ...              ...   
1700                        0                           0                0   
1701                        0                           0                0   
1702                        0                           0                0   
1703                        0                           0                0   
1704                        0                           0                0   

      MedType_Stimulants  MedType_Tranquilizers  SidLvl_mild  ArTyp_DownTown  \
0                      0                      0            0               1   
1                      0                      0            1               0   
2                      0                      0            1               1   
3                      0                      0            1               1   
4                      0                      0            0               0   
...                  ...                    ...          ...             ... 

##### Area_dist_level - Dummy Creation

In [35]:
d = pd.get_dummies(full_data['Area_dist_level'],prefix='ArDisTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_dist_level',axis='columns',inplace = True)

In [36]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

      Availability_rating Area_City_Type  Counterfeit_Sales   Data_from  \
0                0.070422         Tier 1          1775.5026       Train   
1                0.013000         Tier 3          3069.1520       Train   
2                0.060783         Tier 1          2603.0920       Train   
3                0.065555         Tier 1          1101.7130       Train   
4                0.248859         Tier 1           158.9402       Train   
...                   ...            ...                ...         ...   
1700             0.050505         Tier 3                NaN  Production   
1701             0.041118         Tier 2                NaN  Production   
1702             0.191273         Tier 2                NaN  Production   
1703             0.013000         Tier 3                NaN  Production   
1704             0.098982         Tier 2                NaN  Production   

      DistID_Area013  DistID_Area017  ...  MedType_Statins  \
0                  0               0  ...                0   
1                  0               0  ...                0   
2                  0               0  ...                0   
3                  0               0  ...                0   
4                  0               0  ...                0   
...              ...             ...  ...              ...   
1700               0               0  ...                0   
1701               0               0  ...                0   
1702               0               1  ...                0   
1703               0               0  ...                0   
1704               0               0  ...                0   

      MedType_Stimulants  MedType_Tranquilizers  SidLvl_mild  ArTyp_DownTown  \
0                      0                      0            0               1   
1                      0                      0            1               0   
2                      0                      0            1               1   
3                      0                      0            1               1   
4                      0                      0            0               0   
...                  ...                    ...          ...             ...   
1700                   0                      0            1               0   
1701                   0                      0            0               1   
1702                   0                      0            1               1   
1703                   0                      0            1               0   
1704                   0                      0            0               1   

      ArTyp_Industrial  ArTyp_MidTownResidential  ArDisTyp_Medium  \
0                    0                         0                0   
1                    0                         0                1   
2                    0                         0                0   
3                    0                         0                0   
4                    0                         1                0   
...                ...                       ...              ...   
1700                 0                         0                1   
1701                 0       

##### Area_City_Type 	- Dummy Creation

In [37]:
d = pd.get_dummies(full_data['Area_City_Type'],prefix='ArCtyTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_City_Type',axis='columns',inplace = True)

In [38]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  \
0          RRA15              13.100          1995      160.2366   
1          YVV26                 NaN          1983      110.4384   
2          LJC15               9.025          1995      259.4092   
3          GWC40              11.800          1995       99.9830   
4          QMN13                 NaN          1983       56.4402   
...          ...                 ...           ...           ...   
1700       KXW10                 NaN          1983      136.5704   
1701       CKE54              21.300          2002       57.0744   
1702       HAY13              20.400          2005      182.7422   
1703       ZEE32              20.000          2007      266.9672   
1704       NDL93               8.080          2000      238.7694   

      Availability_rating  Counterfeit_Sales   Data_from  DistID_Area013  \
0                0.070422          1775.5026       Train               0   
1                0.013000          3069.1520       Train               0   
2                0.060783          2603.0920       Train               0   
3                0.065555          1101.7130       Train               0   
4                0.248859           158.9402       Train               0   
...                   ...                ...         ...             ...   
1700             0.050505                NaN  Production               0   
1701             0.041118                NaN  Production               0   
1702             0.191273                NaN  Production               0   
1703             0.013000                NaN  Production               0   
1704             0.098982                NaN  Production               0   

      DistID_Area017  DistID_Area018  ...  MedType_Tranquilizers  SidLvl_mild  \
0                  0               0  ...                      0            0   
1                  0               0  ...                      0            1   
2                  0               0  ...                      0            1   
3                  0               0  ...                      0            1   
4                  0               0  ...                      0            0   
...              ...             ...  ...                    ...          ...   
1700               0               0  ...                      0            1   
1701               0               0  ...                      0            0   
1702               1               0  ...                      0            1   
1703               0               1  ...                      0            1   
1704               0               0  ...                      0            0   

      ArTyp_DownTown  ArTyp_Industrial  ArTyp_MidTownResidential  \
0                  1                 0                         0   
1                  0                 0                         0   
2                  1                 0                         0   
3                  1                 0                         0   
4                  0                 0                         1   
...              ...               ...                       ...   
1700               0                 0                         0   
1701               1                 0                         0   
1702               1                 0                         0   
1703               0                 1                         0   
1704               1                 0                         0   

      ArDisTyp_Medium  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  \
0                   0               1                 0                0   
1                   1               0                 0                0   
2                   0               1                 0                0   
3                   0               1                 0                0   
4                   0               1                 0                0   
...               ...        

##### Dropping Medicine ID

In [39]:
full_data.drop('Medicine_ID',axis='columns',inplace=True)

In [40]:
full_data

Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  \
0                 13.100          1995      160.2366             0.070422   
1                    NaN          1983      110.4384             0.013000   
2                  9.025          1995      259.4092             0.060783   
3                 11.800          1995       99.9830             0.065555   
4                    NaN          1983       56.4402             0.248859   
...                  ...           ...           ...                  ...   
1700                 NaN          1983      136.5704             0.050505   
1701              21.300          2002       57.0744             0.041118   
1702              20.400          2005      182.7422             0.191273   
1703              20.000          2007      266.9672             0.013000   
1704               8.080          2000      238.7694             0.098982   

      Counterfeit_Sales   Data_from  DistID_Area013  DistID_Area017  \
0             1775.5026       Train               0               0   
1             3069.1520       Train               0               0   
2             2603.0920       Train               0               0   
3             1101.7130       Train               0               0   
4              158.9402       Train               0               0   
...                 ...         ...             ...             ...   
1700                NaN  Production               0               0   
1701                NaN  Production               0               0   
1702                NaN  Production               0               1   
1703                NaN  Production               0               0   
1704                NaN  Production               0               0   

      DistID_Area018  DistID_Area019  ...  MedType_Tranquilizers  SidLvl_mild  \
0                  0               0  ...                      0            0   
1                  0               0  ...                      0            1   
2                  0               0  ...                      0            1   
3                  0               0  ...                      0            1   
4                  0               1  ...                      0            0   
...              ...             ...  ...                    ...          ...   
1700               0               0  ...                      0            1   
1701               0               0  ...                      0            0   
1702               0               0  ...                      0            1   
1703               1               0  ...                      0            1   
1704               0               0  ...                      0            0   

      ArTyp_DownTown  ArTyp_Industrial  ArTyp_MidTownResidential  \
0                  1                 0                         0   
1                  0                 0                         0   
2                  1                 0                         0   
3                  1                 0                         0   
4                  0                 0                         1   
...              ...               ...                       ...   
1700               0                 0                         0   
1701               1                 0                         0   
1702               1                 0                         0   
1703               0                 1                         0   
1704               1                 0                         0   

      ArDisTyp_Medium  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  \
0                   0               1                 0                0   
1                   1               0                 0                0   
2                   0               1                 0                0   
3                   0               1                 0                0   
4                   0               1                 0   

##### Imputing Counterfeit_weight missing values with mean from train dataset only

In [41]:
wgt_mean = round(full_data.loc[full_data['Data_from'] == 'Train','Counterfeit_Weight'].mean(),3)


In [42]:
wgt_mean

14.115

In [43]:
full_data['Counterfeit_Weight'].fillna(wgt_mean,inplace=True)

In [44]:
full_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 1704
Data columns (total 39 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Counterfeit_Weight          8523 non-null   float64
 1   Active_Since                8523 non-null   int64  
 2   Medicine_MRP                8523 non-null   float64
 3   Availability_rating         8523 non-null   float64
 4   Counterfeit_Sales           6818 non-null   float64
 5   Data_from                   8523 non-null   object 
 6   DistID_Area013              8523 non-null   uint8  
 7   DistID_Area017              8523 non-null   uint8  
 8   DistID_Area018              8523 non-null   uint8  
 9   DistID_Area019              8523 non-null   uint8  
 10  DistID_Area027              8523 non-null   uint8  
 11  DistID_Area035              8523 non-null   uint8  
 12  DistID_Area045              8523 non-null   uint8  
 13  DistID_Area046              8523 

##### Now Saeperating data into Training and Production

In [45]:
training_data = full_data.loc[full_data['Data_from'] == 'Train']
training_data.drop('Data_from',axis=1,inplace=True)

/Users/santu/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [46]:
training_data

Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  \
0                 13.100          1995      160.2366             0.070422   
1                 14.115          1983      110.4384             0.013000   
2                  9.025          1995      259.4092             0.060783   
3                 11.800          1995       99.9830             0.065555   
4                 14.115          1983       56.4402             0.248859   
...                  ...           ...           ...                  ...   
6813               8.535          1995      204.1452             0.112963   
6814              20.650          1995      235.1088             0.131103   
6815              20.000          2005      193.6292             0.105096   
6816              10.180          2000      162.8682             0.099957   
6817               7.930          2007      177.5500             0.117445   

      Counterfeit_Sales  DistID_Area013  DistID_Area017  DistID_Area018  \
0             1775.5026               0               0               0   
1             3069.1520               0               0               0   
2             2603.0920               0               0               0   
3             1101.7130               0               0               0   
4              158.9402               0               0               0   
...                 ...             ...             ...             ...   
6813          2070.4520               0               0               0   
6814          2126.3792               0               0               0   
6815          2119.7212               0               1               0   
6816          1485.2138               0               0               0   
6817          1943.9500               0               0               1   

      DistID_Area019  DistID_Area027  ...  MedType_Tranquilizers  SidLvl_mild  \
0                  0               0  ...                      0            0   
1                  0               1  ...                      0            1   
2                  0               0  ...                      0            1   
3                  0               0  ...                      0            1   
4                  1               0  ...                      0            0   
...              ...             ...  ...                    ...          ...   
6813               0               0  ...                      0            1   
6814               0               0  ...                      0            1   
6815               0               0  ...                      0            0   
6816               0               0  ...                      0            1   
6817               0               0  ...                      0            0   

      ArTyp_DownTown  ArTyp_Industrial  ArTyp_MidTownResidential  \
0                  1                 0                         0   
1                  0                 0                         0   
2                  1                 0                         0   
3                  1                 0                         0   
4                  0                 0                         1   
...              ...               ...                       ...   
6813               1                 0                         0   
6814               1                 0                         0   
6815               1                 0                         0   
6816               1                 0                         0   
6817               0                 1                         0   

      ArDisTyp_Medium  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  \
0                   0               1                 0                0   
1                   1               0                 0                0   
2                   0               1                 0                0   
3                   0               1                 0                0   
4         

In [47]:
production_data = full_data.loc[full_data['Data_from'] == 'Production']

In [48]:
production_data.drop(['Counterfeit_Sales','Data_from'],axis=True,inplace=True)
production_data

/Users/santu/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  \
0                 14.115          1983       85.5328             0.112747   
1                 13.450          2000      257.1460             0.144446   
2                  7.100          2000       98.1172             0.144221   
3                 18.300          1996      135.3730             0.100388   
4                 14.115          1983      112.8016             0.022585   
...                  ...           ...           ...                  ...   
1700              14.115          1983      136.5704             0.050505   
1701              21.300          2002       57.0744             0.041118   
1702              20.400          2005      182.7422             0.191273   
1703              20.000          2007      266.9672             0.013000   
1704               8.080          2000      238.7694             0.098982   

      DistID_Area013  DistID_Area017  DistID_Area018  DistID_Area019  \
0                  0               0               0               0   
1                  0               0               0               0   
2                  0               0               0               0   
3                  0               0               0               0   
4                  0               0               0               1   
...              ...             ...             ...             ...   
1700               0               0               0               0   
1701               0               0               0               0   
1702               0               1               0               0   
1703               0               0               1               0   
1704               0               0               0               0   

      DistID_Area027  DistID_Area035  ...  MedType_Tranquilizers  SidLvl_mild  \
0                  1               0  ...                      0            1   
1                  0               0  ...                      0            1   
2                  0               0  ...                      0            1   
3                  0               0  ...                      1            1   
4                  0               0  ...                      0            1   
...              ...             ...  ...                    ...          ...   
1700               1               0  ...                      0            1   
1701               0               1  ...                      0            0   
1702               0               0  ...                      0            1   
1703               0               0  ...                      0            1   
1704               0               0  ...                      0            0   

      ArTyp_DownTown  ArTyp_Industrial  ArTyp_MidTownResidential  \
0                  0                 0                         0   
1                  1                 0                         0   
2                  1                 0                         0   
3                  0                 0                         1   
4                  0                 0                         1   
...              ...               ...                       ...   
1700               0                 0                         0   
1701               1                 0                         0   
1702               1                 0                         0   
1703               0                 1                         0   
1704               1                 0                         0   

      ArDisTyp_Medium  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  \
0                   1               0                 0                0   
1                   0               0                 1                1   
2                   0               0                 1                1   
3                   0               0                 1                0   
4                   0               1         

# Splitting training data into Train and Test

In [49]:
from sklearn.model_selection import train_test_split

In [50]:
train,test = train_test_split(training_data,test_size=0.2,random_state = 2,)

In [51]:
x_train = train.drop('Counterfeit_Sales',axis='columns')
y_train = train['Counterfeit_Sales']

x_test = test.drop('Counterfeit_Sales',axis='columns')
y_test = test['Counterfeit_Sales']

print('x_train : ',x_train.shape)
print('y_train : ',y_train.shape)

print('x_test : ',x_test.shape)
print('y_test : ',y_test.shape)

x_train :  (5454, 37)
y_train :  (5454,)
x_test :  (1364, 37)
y_test :  (1364,)


# Gradient Boosting

In [52]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn import metrics
from sklearn.model_selection import RandomizedSearchCV

In [53]:
def my_rmse(actual,predicted):
    """Function to calculate Root Mean Square Error"""
    
    return (((actual-predicted)**2).mean())**1/2

In [54]:
x_train.shape

(5454, 37)

In [85]:
params = {
    "n_estimators" : [200,400,600,800,1000],
    'learning_rate' : [0.01,0.05,0.10,0.15,0.20],
    "max_depth" : list(range(3,21)),
    'max_features' : [5,10,15,20,25,30,35,37],
    'subsample' : [0.3,0.4,0.5,0.6,0.7,0.8]
}
iterations = 1000

In [86]:
gbm = GradientBoostingRegressor(random_state=3,verbose=2)

In [87]:
random_gbm = RandomizedSearchCV(gbm,param_distributions=params,cv=5,n_iter=iterations,
                               random_state=2,scoring='neg_mean_absolute_error',n_jobs=-1)

In [88]:
random_gbm.fit(x_train,y_train)

      Iter       Train Loss      OOB Improve   Remaining Time 
         1     2863015.7232       28925.6817            4.24s
         2     2799428.0693       26360.1881            4.71s
         3     2773751.7987       29428.7206            4.80s
         4     2730927.3302       27829.0385            4.94s
         5     2751177.9826       27475.5037            4.88s
         6     2670621.3319       26878.7015            4.87s
         7     2680735.3819       27496.6155            4.81s
         8     2604887.0515       26405.4209            4.83s
         9     2642582.2144       24671.6780            4.77s
        10     2633819.1696       24244.5584            4.78s
        11     2539730.3988       27713.9335            4.78s
        12     2497591.7453       25328.7377            4.78s
        13     2499863.4955       23850.8552            4.71s
        14     2481552.2919       24378.3952            4.72s
        15     2540836.9047       22337.6594            4.66s
       

       137     1257612.3102        3162.9621            3.95s
       138     1269291.9167        2736.9525            3.94s
       139     1268963.3531        2979.3474            3.94s
       140     1278171.4751        2609.6599            3.94s
       141     1252112.9867        3074.7259            3.93s
       142     1270157.2603        3060.0564            3.93s
       143     1295578.3879        2305.6438            3.92s
       144     1285745.2182        1817.1007            3.91s
       145     1268252.6647        2082.5310            3.91s
       146     1292062.2253        2226.1958            3.90s
       147     1261480.7197        2380.5141            3.90s
       148     1257518.6680        2630.7610            3.90s
       149     1282937.4913        2254.2787            3.90s
       150     1265903.4988        2208.4389            3.89s
       151     1291219.4931        2211.1618            3.88s
       152     1273982.4721        2156.5708            3.87s
       1

       276     1126557.9233         221.4559            3.09s
       277     1131095.1677         430.3662            3.08s
       278     1090760.1860          44.4777            3.07s
       279     1146142.7163         138.1546            3.07s
       280     1084817.0660         254.0335            3.06s
       281     1103168.4120         -10.5714            3.06s
       282     1136637.9516         117.8088            3.05s
       283     1101354.8006         178.9827            3.04s
       284     1105525.6316         184.5660            3.04s
       285     1101885.0780        -115.0668            3.03s
       286     1121762.9461         185.0679            3.02s
       287     1127788.1416         173.2500            3.01s
       288     1126430.4249         135.1916            3.01s
       289     1107139.7683          82.0890            3.00s
       290     1078928.7431         185.3247            3.00s
       291     1105471.4523         283.7999            2.99s
       2

       424     1055638.7114        -326.0967            2.16s
       425     1062005.1048          20.5610            2.16s
       426     1048869.0484          22.5582            2.15s
       427     1084775.6036          69.6384            2.14s
       428     1060014.4301         -30.0755            2.14s
       429     1043841.3908        -150.5965            2.13s
       430     1070155.9521        -157.3479            2.13s
       431     1077220.9047        -157.7212            2.12s
       432     1078208.6435         -52.2735            2.11s
       433     1061094.9911          -2.3036            2.11s
       434     1083589.3445         -88.6922            2.10s
       435     1099486.3558         -96.3912            2.10s
       436     1076581.9112          -1.7562            2.09s
       437     1048455.3245          21.1745            2.08s
       438     1072310.2443         -83.6522            2.08s
       439     1058842.2793         -19.4149            2.07s
       4

       572     1054495.0810        -189.2658            1.30s
       573     1030958.0921          49.9109            1.29s
       574     1036088.0471        -138.2606            1.29s
       575     1053226.7438        -109.0142            1.28s
       576     1040380.6755        -219.2282            1.27s
       577     1021528.3574        -385.9425            1.27s
       578     1055949.3458        -112.3372            1.26s
       579     1045248.8661        -159.5475            1.26s
       580     1031367.3082        -197.1772            1.25s
       581     1046760.2100        -178.8499            1.24s
       582     1045515.4292        -109.1547            1.24s
       583     1036673.1584        -265.7384            1.23s
       584     1037665.7610        -189.6806            1.23s
       585     1024098.2597         -40.5182            1.22s
       586     1043134.7433        -508.6615            1.21s
       587     1062061.5911         -74.3331            1.21s
       5

       725      984967.1636         -87.2781            0.42s
       726      999285.9042        -315.6366            0.41s
       727     1027675.4856         -28.6727            0.41s
       728     1022389.1311        -149.0864            0.40s
       729     1029857.1743         -77.6860            0.40s
       730      998408.6774         115.7841            0.39s
       731     1003819.0050        -287.9619            0.39s
       732     1007286.7092         -95.7145            0.38s
       733     1029191.6449        -123.6682            0.37s
       734     1018016.2847        -150.2557            0.37s
       735     1028287.7975         -57.9357            0.36s
       736     1008385.4201        -190.8897            0.36s
       737     1012194.5846         -27.0494            0.35s
       738      979968.3647        -299.2248            0.35s
       739     1033598.1338         -86.6087            0.34s
       740     1032885.1748        -287.5676            0.34s
       7

RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0,
                                                       criterion='friedman_mse',
                                                       init=None,
                                                       learning_rate=0.1,
                                                       loss='ls', max_depth=3,
                                                       max_features=None,
                                                       max_leaf_nodes=None,
                                                       min_impurity_decrease=0.0,
                                                       min_impurity_split=None,
                                                       min_samples_leaf=1,
                                                       min_samples_split=2,
                                                       min_weight_fraction_leaf=0.0,
                             

In [89]:
random_gbm.best_estimator_

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.01, loss='ls', max_depth=3,
                          max_features=30, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=800,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=3, subsample=0.8, tol=0.0001,
                          validation_fraction=0.1, verbose=2, warm_start=False)

In [90]:
random_gbm.cv_results_['mean_test_score']

array([ -821.24962855,  -903.49534527,  -854.65790406,  -827.40155531,
        -844.60936135,  -819.44689379,  -773.87915361,  -801.99202915,
        -977.73494612,  -842.72141679,  -864.01739106,  -883.20746227,
        -815.42876245,  -811.07705606,  -843.69379034,  -875.52772514,
        -852.62945312,  -815.90617262,  -832.32924231,  -812.86127941,
        -850.9081097 ,  -779.41768848,  -843.88516069,  -824.09461733,
        -842.62011368,  -890.39981807,  -870.9030768 ,  -751.94541484,
        -749.52256997,  -857.3390901 ,  -821.44601411,  -815.74104536,
        -756.78115662,  -897.96349731,  -808.80056868,  -822.79775734,
        -857.34300183,  -820.59850032,  -777.91297266,  -797.30500258,
        -831.70751892,  -849.13608417,  -831.13873451,  -994.39122285,
        -811.98466068,  -872.13396623,  -806.53874208,  -853.518474  ,
        -838.59155544,  -866.55584908,  -803.96586225,  -873.36021956,
        -854.57191581,  -858.21151813,  -805.13688252,  -815.56682952,
      

In [91]:
random_gbm.best_score_

-749.5225699663233

In [92]:
random_gbm.best_params_

{'subsample': 0.8,
 'n_estimators': 800,
 'max_features': 30,
 'max_depth': 3,
 'learning_rate': 0.01}

In [93]:
gbm = random_gbm.best_estimator_

In [94]:
gbm.fit(x_train,y_train)
pred = gbm.predict(x_test)

      Iter       Train Loss      OOB Improve   Remaining Time 
         1     2863015.7232       28925.6817            4.57s
         2     2799428.0693       26360.1881            5.16s
         3     2773751.7987       29428.7206            5.14s
         4     2730927.3302       27829.0385            5.01s
         5     2751177.9826       27475.5037            5.25s
         6     2670621.3319       26878.7015            5.32s
         7     2680735.3819       27496.6155            5.29s
         8     2604887.0515       26405.4209            5.30s
         9     2642582.2144       24671.6780            5.22s
        10     2633819.1696       24244.5584            5.17s
        11     2539730.3988       27713.9335            5.10s
        12     2497591.7453       25328.7377            5.04s
        13     2499863.4955       23850.8552            4.92s
        14     2481552.2919       24378.3952            4.89s
        15     2540836.9047       22337.6594            4.79s
       

       167     1245721.8724        1943.2427            3.06s
       168     1193472.2296        1848.4621            3.05s
       169     1211808.5352        1592.9023            3.05s
       170     1198092.3228        1687.1346            3.05s
       171     1179461.4727        1896.2818            3.04s
       172     1201634.6498        1795.5985            3.04s
       173     1202132.6996        1472.2902            3.03s
       174     1178181.2317        1817.1734            3.03s
       175     1214578.0188        1393.5292            3.02s
       176     1238445.7219        1110.9593            3.02s
       177     1201833.9277        1333.4808            3.01s
       178     1200977.4012        1645.9513            3.01s
       179     1240687.8958        1026.3832            3.00s
       180     1230257.3326        1061.1114            3.00s
       181     1209981.7762        1518.1255            2.99s
       182     1187215.3265        1398.4886            2.99s
       1

       335     1078508.1895         255.3922            2.24s
       336     1081461.3507        -218.2385            2.24s
       337     1118220.3607        -160.3244            2.23s
       338     1103001.6354        -175.1334            2.23s
       339     1068229.4356         -22.9252            2.22s
       340     1112363.7120         123.0742            2.22s
       341     1064917.9055          34.6469            2.21s
       342     1042441.9525         106.6009            2.21s
       343     1066388.3666         127.0278            2.21s
       344     1086050.8839          15.4076            2.20s
       345     1107010.5449          77.8013            2.20s
       346     1090752.5539          49.6166            2.19s
       347     1040406.7723          44.2094            2.19s
       348     1072375.8200         -79.7489            2.18s
       349     1089675.8274          39.1529            2.18s
       350     1105448.6499         -23.4484            2.17s
       3

       493     1039761.3588         -51.8055            1.51s
       494     1050411.5167        -169.7280            1.51s
       495     1043980.6118        -160.0266            1.50s
       496     1035600.4634        -218.6181            1.50s
       497     1081632.8870         -67.0878            1.49s
       498     1049912.2834         -97.8920            1.49s
       499     1064201.7648         -20.0644            1.48s
       500     1048112.6091          31.9384            1.48s
       501     1054238.0903        -216.0343            1.47s
       502     1057311.4547         -69.0744            1.47s
       503     1061901.6840          41.0467            1.46s
       504     1064373.7011         -70.9598            1.46s
       505     1050497.6103         -31.7641            1.45s
       506     1055438.7058         -90.6993            1.45s
       507     1048470.7979        -136.9907            1.44s
       508     1031712.2410        -143.0645            1.44s
       5

       669     1052750.3100        -108.3693            0.63s
       670     1020019.6700        -158.4451            0.63s
       671     1007944.5426         -47.1682            0.62s
       672     1031384.1834         -61.7666            0.62s
       673     1048481.3488        -463.8818            0.61s
       674     1005812.1569         -70.0413            0.61s
       675     1047002.7509        -298.1942            0.60s
       676     1053667.5271         -92.4289            0.60s
       677     1025627.9484          -5.6026            0.59s
       678     1028124.2429         -19.9166            0.59s
       679     1041559.0105          -1.1592            0.59s
       680     1006417.4625        -120.4761            0.58s
       681     1023324.7813        -212.7420            0.58s
       682     1010318.8527         -81.2713            0.57s
       683      985955.7815        -400.7176            0.57s
       684     1018878.6671         -34.3885            0.56s
       6

In [95]:
print('performance on test data')
print('Mean Absolute Error = ', metrics.median_absolute_error(y_test,pred))
print('Root Mean Square Error = ',my_rmse(y_test,pred))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_test,pred)/1600))

performance on test data
Mean Absolute Error =  481.02319571185
Root Mean Square Error =  604809.0510707761
Our model submission performance :  0.6993605026800938


In [96]:
print("performance on train data")
pred_train = gbm.predict(x_train)
print('Mean Absolute Error = ', metrics.median_absolute_error(y_train,pred_train))
print('Root Mean Square Error = ',my_rmse(y_train,pred_train))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_train,pred_train)/1600))

performance on train data
Mean Absolute Error =  485.3583834521239
Root Mean Square Error =  503901.48446975095
Our model submission performance :  0.6966510103424226


##### Building model on full training data and submission

In [ ]:
x_training = training_data.drop('Counterfeit_Sales',axis='columns')
y_training = training_data['Counterfeit_Sales']

In [ ]:
lr.fit(x_training,y_training)
pred = lr.predict(production_data)

In [ ]:
production_medID = pd.read_csv(file_path + 'counterfeit_test.csv')

In [ ]:
production_medID = production_medID['Medicine_ID']

In [ ]:
submission_df = pd.DataFrame({"Medicine_ID": production_medID,
                 'Counterfeit_Sales': pred})

In [ ]:
submission_df.to_csv('part2 project3.csv',index=False)

# Conclusion

### Linear Regression

Mean Absolute Error =  615.9795920660683

Root Mean Square Error =  641656.7346209903

Our model submission performance :  0.6150127549587072

### Decision Tree

Mean Absolute Error =  488.0314000000002

Root Mean Square Error =  625993.4594418913

Our model submission performance :  0.6949803749999999


### Random Forest

performance on test data

Mean Absolute Error =  491.96711024999945

Root Mean Square Error =  622689.0509111757

Our model submission performance :  0.6925205560937504

### Extra Trees

performance on test data

Mean Absolute Error =  473.6709262499995

Root Mean Square Error =  618246.3458160626

Our model submission performance :  0.7039556710937502

### Gradient Boost Machine

performance on test data

Mean Absolute Error =  481.02319571185

Root Mean Square Error =  604809.0510707761

Our model submission performance :  0.6993605026800938
